packages/jellyfish-transaction/src/buffer/buffer_composer.ts

```diff
@@ -209,6 +209,32 @@ export abstract class ComposableBuffer<T> implements BufferComposer {
     }
   }
 
+  /**
+   *  HEX String without length specified, encoded into Buffer as the same order of the Hex String.
+   *  In short this read a hex and push it into the Buffer. It will not re-order the endian.
+   * `toBuffer` defaults to '' and writes 00 to buffer
+   *
+   * @param getter to read HEX String. Writes its length then write the HEX string. Defaults to empty string.
+   * @param setter to read ordered Buffer and set as the same ordered HEX String
+   */
+  static optionalVarUIntHex (getter: () => string | undefined, setter: (data: string | undefined) => void): BufferComposer {
+    return {
+      fromBuffer: (buffer: SmartBuffer): void => {
+        const length = readVarUInt(buffer)
+        if (length > 0) {
```

Comment:
Here, is this correct?
the first byte of the pubkey is not the length, but more of an indicator of length. i.e compressed or uncompressed pubkey.
Refer https://github.com/DeFiCh/ain/blob/e6caba0c0b220182b69e6043502a51c3f567c599/src/pubkey.h#L57
and the `CPubKey` class

Reply:
The check is here to handle optional property.
If it was not passed, length will equal 0, otherwise it means the pubkey is present

Reply:
no not that. i'm talking more of the overall functionality of `optionalVarUIntHex`. It's incorrect.
ok below is a legit ICX createOrder tx scriptPubKeys
```
6a 4c 5d 44665478 31 01 0017a9140125 8bd2b4d6409f0c0c831cfe7b75d6fe79da708721 037f9563f30c609b19fd435a19b8bde7d6db703012ba1aba72e9f42a87366d1941 002f685900000000 002f685900000000 40420f0000000000 400b0000

6a 3c 44665478 31 02 0017a9149d2f 179fc782f4250fe59696f6319e1391b705ee8700 00c2eb0b00000000 00c2eb0b00000000 00e8764817000000 400b0000
```

The first one for INTERNAL type, second one is for EXTERNAL. I've separated them by space for easy understanding.
test here https://github.com/DeFiCh/jellyfish/pull/562/files/9c5234cca2035cadfc7fee9c390f8277f49f4164#r684171280 are incorrect. that's why it was not failed.
hope it's clear now.

Reply:
6a 4c 5d 44665478 31 01 0017a9140125 8bd2b4d6409f0c0c831cfe7b75d6fe79da708721 037f9563f30c609b19fd435a19b8bde7d6db703012ba1aba72e9f42a87366d1941 002f685900000000 002f685900000000 40420f0000000000 400b0000

6a 3c 44665478 31 02 0017a9149d2f 179fc782f4250fe59696f6319e1391b705ee8700 00c2eb0b00000000 00c2eb0b00000000 00e8764817000000 400b0000


6a 4c 5d 44665478 31 01 0017a9140125 8bd2b4d6409f0c0c831cfe7b75d6fe79da7087 **21** 037f9563f30c609b19fd435a19b8bde7d6db703012ba1aba72e9f42a87366d1941 002f685900000000 002f685900000000 40420f0000000000 400b0000

6a 3c 44665478 31 02 0017a9149d2f 179fc782f4250fe59696f6319e1391b705ee87 **00** 00c2eb0b00000000 00c2eb0b00000000 00e8764817000000 400b0000

From what I tested, these values represent the length of the receivePubkey, 00 representing an absence.
The ownerAddress ends with 0x87 OP_EQUAL right before this length

Reply:
I'm incorrect. `CScript` and `CPubKey` were not serialized the way I imagined. following is the correct separation of items from scriptPubKey
```
6a 4c 5d 44665478 31 01 00 17a91401258bd2b4d6409f0c0c831cfe7b75d6fe79da7087 21037f9563f30c609b19fd435a19b8bde7d6db703012ba1aba72e9f42a87366d1941 002f685900000000 002f685900000000 40420f0000000000 400b0000
6a 3c 44665478 31 02 00 17a9149d2f179fc782f4250fe59696f6319e1391b705ee87 00 00c2eb0b00000000 00c2eb0b00000000 00e8764817000000 400b0000
```
The first one for INTERNAL type order, second one is for EXTERNAL order
note that in our case, `DCT_ID` only takes one byte.
so your `optionalVarUIntHex()` is correct :clap:

But change the order type to `Internal` in here https://github.com/DeFiCh/jellyfish/pull/562/files/9c5234cca2035cadfc7fee9c390f8277f49f4164#r684171280
coz for external types, `receivePubkey` does not matter anyways

Reply:
Ok I will test Internal with receivePubkey and External without receivePubkey

Reply:
2bfb4d2